File: 239-sliding-window-maximum/239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        res = [0] * (n - k + 1)
        q = deque()
        
        for i in range(n):
            idx = i - k + 1
            while q and i - q[0] + 1 > k:
                q.popleft()
            while q and nums[i] >= nums[q[-1]]:
                q.pop()
            q.append(i)
            if idx >= 0:
                res[idx] = nums[q[0]]
        return res

File: 239-sliding-window-maximum/239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] res = new int[n - k + 1];
        Deque<Integer> q = new ArrayDeque<>();
        for(int i = 0; i < n; i++) {
            int idx = i - k + 1;
            //取出左边界元素
            while(!q.isEmpty() && i - q.peekFirst()  + 1 > k) {
                q.pollFirst();
            }
            //保持queue的单点递减性
            while(!q.isEmpty() && nums[i] >= nums[q.peekLast()]) {
                q.pollLast();
            }
            q.offerLast(i);
            if(idx >= 0) res[idx] = nums[q.peekFirst()];
        }
        return res;
    }
}